File: 03_Combinatorics_Search/20_Combination/base_code.py
def generate_combinations(arr, r):
    """
    백트래킹을 이용하여 arr에서 순서에 상관없이 r개를 뽑는 모든 조합을 구합니다.
    """
    result = []
    
    # start: 이번 재귀에서 탐색을 시작할 인덱스 (이전 원소는 쳐다보지 않음)
    def backtrack(start, path):
        # 1. 종료 조건: 원하는 개수(r)만큼 뽑았다면 결과에 추가
        if len(path) == r:
            result.append(path[:])
            return
        
        # 2. 탐색: start 인덱스부터 배열의 끝까지 확인
        for i in range(start, len(arr)):
            path.append(arr[i])         # 현재 원소를 뽑음
            
            # 다음 원소는 현재 뽑은 원소의 '다음 인덱스(i + 1)'부터 탐색하도록 넘김
            backtrack(i + 1, path)      
            
            # 원래 상태로 복구 (백트래킹)
            path.pop()                  
                
    backtrack(0, [])
    return result

# 사용 예시
if __name__ == "__main__":
    data = [1, 2, 3, 4]
    r = 2
    
    # 4개 중 2개를 뽑는 조합 (4C2 = 6개)
    print(f"{data}에서 {r}개를 뽑는 조합:")
    for c in generate_combinations(data, r):
        print(c)

File: 03_Combinatorics_Search/20_Combination/BaseCode.java
import java.util.Arrays;

public class BaseCode {

    /**
     * arr 배열에서 r개를 뽑는 조합을 생성하는 메인 헬퍼 함수
     */
    public static void generateCombinations(int[] arr, int r) {
        int[] out = new int[r]; // 뽑힌 원소를 순서대로 담을 배열
        
        // start: 0번 인덱스부터 탐색 시작, depth: 현재 0개 뽑음
        backtrack(arr, out, 0, 0, r);
    }

    /**
     * 백트래킹을 이용한 조합 재귀 함수
     * @param start 탐색을 시작할 배열의 인덱스 (중복 선택 및 역방향 탐색 방지)
     * @param depth 현재까지 뽑은 원소의 개수
     */
    private static void backtrack(int[] arr, int[] out, int start, int depth, int r) {
        // 1. 종료 조건: 원하는 개수(r)만큼 다 뽑았다면 출력
        if (depth == r) {
            System.out.println(Arrays.toString(out));
            return;
        }

        // 2. 탐색: start부터 끝까지 확인 (이전 인덱스는 무시)
        for (int i = start; i < arr.length; i++) {
            out[depth] = arr[i]; // 현재 깊이에 원소 저장
            
            // 3. 다음 원소를 뽑기 위해 재귀 호출 (다음 탐색은 i + 1부터)
            backtrack(arr, out, i + 1, depth + 1, r);
        }
    }

    // 사용 예시
    public static void main(String[] args) {
        int[] data = {1, 2, 3, 4};
        int r = 2; // 4개 중 2개를 뽑는 조합
        
        System.out.println("자바 백트래킹을 이용한 조합:");
        generateCombinations(data, r);
    }
}